R/calculateCoex-method.R

```diff
@@ -873,85 +1168,49 @@ contingencyTables <- function(objCOTAN, g1, g2) {
-
-    gc()
-
-    logThis(paste("Calculating", kind, "coex normalization factor"),
-            logLevel = 3L)
+  function(objCOTAN, actOnCells = FALSE, returnPPFract = FALSE,
+           optimizeForSpeed = TRUE, deviceStr = "cuda") {
+    coex <- NULL
+    problematicPairsFraction <- NA
 
     if (isTRUE(actOnCells)) {
-      allNames <- getCells(objCOTAN)
-      normFact <- 1.0 / sqrt(getNumGenes(objCOTAN)) # divided by sqrt(n)
-    } else {
-      allNames <- getGenes(objCOTAN)
-      normFact <- 1.0 / sqrt(getNumCells(objCOTAN)) # divided by sqrt(m)
-    }
-
-    problematicPairsFraction <-
-      sum(pmin(expectedYY@x, expectedYN@x,
-               expectedNY@x, expectedNN@x) < 0.5) / length(expectedYY@x)
-
-    logThis(paste0("Fraction of genes with very low",
-                   " expected contingency tables: ",
-                   problematicPairsFraction), logLevel = 3L)
-
-    coex <- normFact * sqrt(1.0 / pmax(1.0, expectedYY@x) +
-                            1.0 / pmax(1.0, expectedYN@x) +
-                            1.0 / pmax(1.0, expectedNY@x) +
-                            1.0 / pmax(1.0, expectedNN@x))
-
-    rm(expectedYN, expectedNY, expectedNN)
-    gc()
+      if(isTRUE(optimizeForSpeed)) {
+        warning("The 'torch' package is not supported yet for cells' COEX")
+      }
+      c(coex, problematicPairsFraction) %<-%
+        calculateCoex_Legacy(objCOTAN, actOnCells = TRUE,
+                             returnPPFract = returnPPFract)
 
-    logThis(paste("Retrieving observed", kind, "yes/yes contingency table"),
-            logLevel = 3L)
-
-    c(observedYY, .) %<-%
-      observedContingencyTablesYY(objCOTAN,
-                                  actOnCells = actOnCells,
-                                  asDspMatrices = TRUE)
-
-    gc()
-
-    # coex estimation
-    logThis(paste("Estimating", kind, "coex"), logLevel = 3L)
-
-    coex <- coex * (observedYY@x - expectedYY@x)
-
-    assert_that(2L * length(coex) == length(allNames) * (length(allNames) + 1L),
-                msg = "Output coex@x has the wrong size")
-
-    coex <- new("dspMatrix", Dim = dim(expectedYY), x = coex)
-    rownames(coex) <- colnames(coex) <- allNames
-
-    rm(observedYY, expectedYY)
-    gc()
-
-    if (actOnCells) {
       objCOTAN@cellsCoex <- coex
       objCOTAN@metaDataset <- updateMetaInfo(objCOTAN@metaDataset,
                                              datasetTags()[["csync"]], TRUE)
       objCOTAN@metaDataset <- updateMetaInfo(objCOTAN@metaDataset,
                                              datasetTags()[["cbad"]],
                                              problematicPairsFraction)
     } else {
+      if (optimizeForSpeed && requireNamespace("torch", quietly = TRUE)) {
+        library("torch", character.only = TRUE)
+
+        # Device configuration - fall-back to cpu if no cuda device is available
+        deviceStr <- ifelse(deviceStr == "cuda" && torch::cuda_is_available(),
```

Comment:
Does this work if deviceStr is "cuda:1" or similar?

Reply:
Ops...
Correggo...